pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -3317,131 +3317,184 @@ protected void internalTerminatePartitionedTopic(AsyncResponse asyncResponse, bo
+                                                    expireTimeInSeconds, topicName, cause);
+                                            resumeAsyncResponseExceptionally(asyncResponse, cause);
+                                            return null;
+                                        });
+                            }
                         }
-                    }
+                    }).exceptionally(e -> {
+                        Throwable cause = e.getCause();
+                        log.error("[{}] Failed to expire messages up to {} on {}", clientAppId(), expireTimeInSeconds,
+                                topicName, cause);
+                        resumeAsyncResponseExceptionally(asyncResponse, cause);
+                        return null;
+                    })).exceptionally(e -> {
+                        Throwable cause = e.getCause();
+                        log.error("[{}] Failed to validate global namespace ownership to expire messages up to {} on {}"
+                                , clientAppId(), expireTimeInSeconds, topicName, cause);
+                        resumeAsyncResponseExceptionally(asyncResponse, cause);
+                        return null;
+                    });
+    }
 
-                    asyncResponse.resume(Response.noContent().build());
-                    return null;
-                });
+    // this method used in internalExpireMessagesForAllSubscriptionsForNonPartitionedTopic,
+    private void internalExpireMessagesByTimestampForSinglePartition(String subName, int expireTimeInSeconds,
+            boolean authoritative) {
+        PartitionedTopicMetadata partitionMetadata = getPartitionedTopicMetadata(topicName, authoritative, false);
+        try {
+            internalExpireMessagesByTimestampForSinglePartitionAsync(partitionMetadata, subName, expireTimeInSeconds,
+                    authoritative).join();
+        } catch (CompletionException ce) {
+            if (ce.getCause() instanceof WebApplicationException) {
+                throw (WebApplicationException) ce.getCause();
             } else {
-                try {
-                    internalExpireMessagesByTimestampForSinglePartition(subName, expireTimeInSeconds, authoritative);
-                } catch (WebApplicationException wae) {
-                    asyncResponse.resume(wae);
-                    return;
-                } catch (Exception e) {
-                    asyncResponse.resume(new RestException(e));
-                    return;
-                }
-                asyncResponse.resume(Response.noContent().build());
+                throw new RestException(ce.getCause());
             }
         }
     }
 
-    private void internalExpireMessagesByTimestampForSinglePartition(String subName, int expireTimeInSeconds,
+    private CompletableFuture<Void> internalExpireMessagesByTimestampForSinglePartitionAsync(
+            PartitionedTopicMetadata partitionMetadata, String subName, int expireTimeInSeconds,
```

Comment:
Changing the method signature may produce greater changes， because both `nternalExpireMessagesForAllSubscriptionsForNonPartitionedTopicAsync` will also invoke it

Reply:
nice catch! @liudezhi2098 Could you please help check all the `internalExpireMessagesForAllSubscriptionsForNonPartitionedTopic` method?

Reply:
OK，I will fix both of these methods in this PR